restart.c

```diff
@@ -0,0 +1,411 @@
+    uint64_t checked = 0;
+    const unsigned int page_size = settings.slab_page_size;
+    unsigned int page_remain = page_size;
+
+    gettimeofday(&tv, NULL);
+    if (settings.verbose > 0) {
+        fprintf(stderr, "[restart] original memory base: [%p] new base: [%p]\n", orig_addr, mmap_base);
+        fprintf(stderr, "[restart] recovery start [%d.%d]\n", (int)tv.tv_sec, (int)tv.tv_usec);
+    }
+
+    // since chunks don't align with pages, we have to also track page size.
+    while (checked < slabmem_limit) {
+        //fprintf(stderr, "checked: %lu\n", checked);
+        item *it = (item *)((char *)mmap_base + checked);
+
+        int size = slabs_fixup((char *)mmap_base + checked,
+                checked % settings.slab_page_size);
+        //fprintf(stderr, "id: %d, size: %d\n", it->slabs_clsid, size);
+        // slabber gobbled an entire page, skip and move on.
+        if (size == -1) {
+            assert(page_remain % page_size == 0);
+            assert(page_remain == page_size);
+            checked += page_remain;
+            page_remain = page_size;
+            continue;
+        }
+
+        if (it->it_flags & ITEM_LINKED) {
+            // fixup next/prev links while on LRU.
+            if (it->next) {
+                it->next = (item *)((uint64_t)it->next - (uint64_t)orig_addr);
+                it->next = (item *)((uint64_t)it->next + (uint64_t)mmap_base);
+            }
+            if (it->prev) {
+                it->prev = (item *)((uint64_t)it->prev - (uint64_t)orig_addr);
+                it->prev = (item *)((uint64_t)it->prev + (uint64_t)mmap_base);
+            }
+
+            //fprintf(stderr, "item was linked\n");
+            do_item_link_fixup(it);
+        }
+
+        if (it->it_flags & (ITEM_CHUNKED|ITEM_CHUNK)) {
+            item_chunk *ch;
+            if (it->it_flags & ITEM_CHUNKED) {
+                ch = (item_chunk *) ITEM_schunk(it);
+                // Sigh. Chunked items are a hack; the clsid is the clsid of
+                // the full object (always the largest slab class) rather than
+                // the actual chunk.
+                // I bet this is fixable :(
+                size = slabs_size(ch->orig_clsid);
+                //fprintf(stderr, "fixing chunked item header [%d]\n", size);
+            } else {
+                //fprintf(stderr, "fixing item chunk [%d]\n", size);
+                ch = (item_chunk *) it;
+            }
+            if (ch->next) {
+                ch->next = (item_chunk *)((uint64_t)ch->next - (uint64_t)orig_addr);
+                ch->next = (item_chunk *)((uint64_t)ch->next + (uint64_t)mmap_base);
+            }
+            if (ch->prev) {
+                ch->prev = (item_chunk *)((uint64_t)ch->prev - (uint64_t)orig_addr);
+                ch->prev = (item_chunk *)((uint64_t)ch->prev + (uint64_t)mmap_base);
+            }
+            if (ch->head) {
+                ch->head = (item *)((uint64_t)it->prev - (uint64_t)orig_addr);
```

Comment:
womp womp. it->prev -> ch->head. how was this passing tests? The slab mover blows up immediately when passing a fixed up item here, so ch->head is probably rarely used.